File: packages/restapi/src/lib/chat/createGroup.ts
import { getAPIBaseUrls } from '../helpers';
import Constants from '../constants';
import { EnvOptionsType, GroupDTO, SignerType, Rules } from '../types';
import {
  ICreateGroupRequestPayload,
  createGroupPayload,
  createGroupRequestValidator,
  getWallet,
  getUserDID,
  IPGPHelper,
  PGPHelper,
  validateScheduleDates,
  getConnectedUserV2Core,
} from './helpers';
import * as CryptoJS from 'crypto-js';
import { axiosPost } from '../utils/axiosUtil';
import { handleError } from '../errors/ValidationError';

export interface ChatCreateGroupType extends EnvOptionsType {
  account?: string | null;
  signer?: SignerType | null;
  groupName: string;
  groupDescription?: string | null;
  members: Array<string>;
  groupImage?: string | null;
  admins: Array<string>;
  isPublic: boolean;
  contractAddressNFT?: string;
  numberOfNFTs?: number;
  contractAddressERC20?: string;
  numberOfERC20?: number;
  pgpPrivateKey?: string | null;
  meta?: string;
  groupType?: string | null;
  scheduleAt?: Date | null;
  scheduleEnd?: Date | null;
  rules?: Rules | null;
}

export const createGroup = async (options: ChatCreateGroupType) => {
  return await createGroupCore(options, PGPHelper);
};

export const createGroupCore = async (
  options: ChatCreateGroupType,
  pgpHelper: IPGPHelper
): Promise<GroupDTO> => {
  const {
    account = null,
    signer = null,
    groupName,
    groupDescription,
    members,
    groupImage,
    admins,
    isPublic,
    contractAddressNFT,
    numberOfNFTs,
    contractAddressERC20,
    numberOfERC20,
    env = Constants.ENV.PROD,
    pgpPrivateKey = null,
    meta,
    groupType,
    scheduleAt,
    scheduleEnd,
    rules,
  } = options || {};

  try {
    if (account == null && signer == null) {
      throw new Error(`At least one from account or signer is necessary!`);
    }

    validateScheduleDates(scheduleAt, scheduleEnd);

    const wallet = getWallet({ account, signer });

    createGroupRequestValidator(
      groupName,
      members,
      admins,
      groupDescription,
      contractAddressNFT,
      numberOfNFTs,
      contractAddressERC20,
      numberOfERC20
    );

    const convertedMembersPromise = members.map(async (each) => {
      return getUserDID(each, env);
    });
    const convertedAdminsPromise = admins.map(async (each) => {
      return getUserDID(each, env);
    });
    const convertedMembers = await Promise.all(convertedMembersPromise);
    const convertedAdmins = await Promise.all(convertedAdminsPromise);

    const connectedUser = await getConnectedUserV2Core(
      wallet,
      pgpPrivateKey,
      env,
      pgpHelper
    );

    const bodyToBeHashed = {
      groupName: groupName,
      groupDescription: groupDescription == undefined ? null : groupDescription,
      members: convertedMembers,
      groupImage: groupImage == undefined ? null : groupImage,
      admins: convertedAdmins,
      isPublic: isPublic,
      contractAddressNFT:
        contractAddressNFT == undefined ? null : contractAddressNFT,
      numberOfNFTs: numberOfNFTs == undefined ? 0 : numberOfNFTs,
      contractAddressERC20:
        contractAddressERC20 == undefined ? null : contractAddressERC20,
      numberOfERC20: numberOfERC20 == undefined ? 0 : numberOfERC20,
      groupCreator: connectedUser.did,
    };

    const hash = CryptoJS.SHA256(JSON.stringify(bodyToBeHashed)).toString();
    const signature: string = await pgpHelper.sign({
      message: hash,
      signingKey: connectedUser.privateKey!,
    });
    const sigType = 'pgp';

    const verificationProof: string = sigType + ':' + signature;

    const API_BASE_URL = getAPIBaseUrls(env);
    const apiEndpoint = `${API_BASE_URL}/v1/chat/groups`;
    const body: ICreateGroupRequestPayload = createGroupPayload(
      groupName,
      convertedMembers,
      convertedAdmins,
      isPublic,
      connectedUser.did,
      verificationProof,
      groupDescription,
      groupImage,
      contractAddressNFT,
      numberOfNFTs,
      contractAddressERC20,
      numberOfERC20,
      meta,
      groupType,
      scheduleAt,
      scheduleEnd,
      rules
    );

    const response = await axiosPost(apiEndpoint, body);
    return response.data;
  } catch (err) {
      throw handleError(err, createGroup.name);
  }
};

File: packages/push-api-dart/lib/src/chat/src/create_group.dart
import '../../../push_api_dart.dart';

createGroup({
  String? account,
  Signer? signer,
  required String groupName,
  required String groupDescription,
  required String groupImage,
  required List<String> members,
  required List<String> admins,
  required bool isPublic,
  String? contractAddressNFT,
  int? numberOfNFTs,
  String? contractAddressERC20,
  int? numberOfERC20,
  String? pgpPrivateKey,
  String? meta,
}) async {
  try {
    if (account == null && signer == null) {
      throw Exception('At least one from account or signer is necessary!');
    }
    final wallet = getWallet(address: account, signer: signer);
    String address = getAccountAddress(wallet);

    createGroupRequestValidator(
        groupName: groupName,
        groupDescription: groupDescription,
        members: members,
        admins: admins);

    final convertedMembersDIDList =
        await Future.wait(members.map((item) => getUserDID(address: item)));
    final convertedAdminsDIDList =
        await Future.wait(admins.map((item) => getUserDID(address: item)));

    final connectedUser = await getConnectedUserV2(
      wallet: wallet,
      privateKey: pgpPrivateKey,
    );

    final bodyToBeHashed = {
      'groupName': groupName,
      'groupDescription': groupDescription,
      'members': convertedMembersDIDList,
      'groupImage': groupImage,
      'admins': convertedAdminsDIDList,
      'isPublic': isPublic,
      'contractAddressNFT': contractAddressNFT,
      'numberOfNFTs': numberOfNFTs ?? 0,
      'contractAddressERC20': contractAddressERC20,
      'numberOfERC20': numberOfERC20 ?? 0,
      'groupCreator': address,
    };

    final hash = generateHash(bodyToBeHashed);

    final signature = await sign(
      message: hash,
      privateKey: connectedUser!.user.encryptedPrivateKey!,
      publicKey: connectedUser.user.publicKey!,
    );

    const sigType = 'pgp';

    final String verificationProof = '$sigType:$signature';

    final body = {
      'groupName': groupName,
      'groupDescription': groupDescription,
      'members': members,
      'groupImage': groupImage,
      'admins': admins,
      'isPublic': isPublic,
      'contractAddressNFT': contractAddressNFT,
      'numberOfNFTs': numberOfNFTs,
      'contractAddressERC20': contractAddressERC20,
      'numberOfERC20': numberOfERC20,
      'groupCreator': address,
      'verificationProof': verificationProof,
      'meta': meta,
    };

    final result = await http.post(
      path: '/v1/chat/groups',
      data: body,
    );

    if (result == null) {
      throw Exception(result);
    }

    return GroupDTO.fromJson(result);
  } catch (e) {
    log("[Push SDK] - API  - Error - API createGroup -: $e ");
    rethrow;
  }
}
